File: app/src/main/java/com/ben/sudoku/SudokuView.java
package com.ben.sudoku;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

/**
 * 数独视图
 *
 * @author chenbenbin
 * @date 2018/7/11
 */
public class SudokuView extends View {
    private Paint mSourcePaint = new Paint();
    private Paint mInputPaint = new Paint();
    private Paint mTipsPaint = new Paint();
    private Paint mBoldLinePaint = new Paint();
    private Paint mThinLinePaint = new Paint();
    private int mRectSize;
    private int mPanelMarginLeft;
    private int mPanelMarginTop;
    private int mTextSize;
    private int mTextBaseline;
    private int[] mNoteTextBaseline = new int[3];
    private SudokuPresenter mSudokuPresenter = new SudokuPresenter();

    public SudokuView(Context context) {
        super(context);
        init();
    }

    public SudokuView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public SudokuView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        mSourcePaint.setAntiAlias(true);
        mSourcePaint.setTextAlign(Paint.Align.CENTER);

        mInputPaint.setTextAlign(Paint.Align.CENTER);
        mInputPaint.setColor(getResources().getColor(R.color.colorPrimary));

        mTipsPaint.setAntiAlias(true);
        mTipsPaint.setStrokeWidth(5);
        mTipsPaint.setTextAlign(Paint.Align.CENTER);

        mBoldLinePaint.setStrokeWidth(5);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);

        mRectSize = w / 10;
        int rectPadding = mRectSize / 5;
        mPanelMarginLeft = mRectSize / 2;
        mPanelMarginTop = (h - w) / 2 + mPanelMarginLeft;
        mTextSize = mRectSize - rectPadding * 2;
        mSourcePaint.setTextSize(mTextSize);

        Paint.FontMetricsInt fontMetrics = mSourcePaint.getFontMetricsInt();
        mTextBaseline = (mRectSize - fontMetrics.bottom + fontMetrics.top) / 2 - fontMetrics.top;

        Paint.FontMetricsInt noteFontMetrics = mTipsPaint.getFontMetricsInt();
        for (int i = 0; i < mNoteTextBaseline.length; i++) {
            mNoteTextBaseline[i] = (mRectSize / 3 * (i * 2 + 1) - noteFontMetrics.bottom + noteFontMetrics.top) / 2 -
                    noteFontMetrics.top;
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int width = canvas.getWidth();
        int height = canvas.getHeight();

        for (int i = 0; i < SudokuPresenter.GRID_COUNT; i++) {
            int column = i % 9;
            int row = i / 9;

            GridBean gridBean = mSudokuPresenter.getGrid(i);
            if (gridBean.getSource() != 0) {
                drawSource(canvas, gridBean.getSource(), column, row);
            } else if (gridBean.getInput() != 0) {
                drawInput(canvas, gridBean.getInput(), column, row);
            } else {
                drawTips(canvas, gridBean.getTips(), column, row);
            }
        }
        for (int i = 0; i < 10; i++) {
            canvas.drawLine(mPanelMarginLeft, mPanelMarginTop + mRectSize * i, width - mPanelMarginLeft, mPanelMarginTop +
                    mRectSize * i, i % 3 == 0 ?
                    mBoldLinePaint : mThinLinePaint);
        }
        for (int i = 0; i < 10; i++) {
            canvas.drawLine(mPanelMarginLeft + mRectSize * i, mPanelMarginTop, mPanelMarginLeft + mRectSize * i, height -
                    mPanelMarginTop, i % 3 == 0 ?
                    mBoldLinePaint : mThinLinePaint);
        }
    }

    /**
     * 绘制原始的数字
     */
    private void drawSource(Canvas canvas, int num, int column, int row) {
        canvas.drawText(String.valueOf(num),
                mPanelMarginLeft + column * mRectSize + mRectSize / 2,
                mPanelMarginTop + mTextBaseline + row * mRectSize,
                mSourcePaint);
    }

    /**
     * 绘制输入的数字
     */
    private void drawInput(Canvas canvas, int num, int column, int row) {
        canvas.drawText(String.valueOf(num),
                mPanelMarginLeft + column * mRectSize + mRectSize / 2,
                mPanelMarginTop + mTextBaseline + row * mRectSize,
                mInputPaint);
    }

    /**
     * 绘制提示数字
     */
    private void drawTips(Canvas canvas, int[] tips, int column, int row) {
        for (int j = 0; j < 9; j++) {
            if (tips[j] == 0) {
                continue;
            }
            canvas.drawText(String.valueOf(j + 1),
                    mPanelMarginLeft + column * mRectSize + mRectSize / 6 * (j % 3 * 2 + 1),
                    mPanelMarginTop + mNoteTextBaseline[j / 3] + row * mRectSize,
                    mTipsPaint);
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mSudokuPresenter.updateTips();
        return super.onTouchEvent(event);
    }
}

File: app/src/main/java/com/ben/sudoku/SudokuView.kt
package com.ben.sudoku

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.widget.Toast

/**
 * 数独视图
 *
 * @author Benhero
 * @date 2018/7/11
 */
class SudokuView : View {
    private val mSourcePaint = Paint()
    private val mInputPaint = Paint()
    private val mTipsPaint = Paint()
    private val mBoldLinePaint = Paint()
    private val mThinLinePaint = Paint()
    private var mRectSize: Int = 0
    private var mPanelMarginLeft: Int = 0
    private var mPanelMarginTop: Int = 0
    private var mTextSize: Int = 0
    private var mTextBaseline: Int = 0
    private val mNoteTextBaseline = IntArray(3)
    private val mSudokuPresenter = SudokuPresenter()

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        mSourcePaint.isAntiAlias = true
        mSourcePaint.textAlign = Paint.Align.CENTER

        mInputPaint.isAntiAlias = true
        mInputPaint.textAlign = Paint.Align.CENTER
        mInputPaint.color = resources.getColor(R.color.colorPrimary)

        mTipsPaint.isAntiAlias = true
        mTipsPaint.strokeWidth = 5f
        mTipsPaint.textAlign = Paint.Align.CENTER

        mBoldLinePaint.strokeWidth = 5f
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        mRectSize = w / 10
        val rectPadding = mRectSize / 5
        mPanelMarginLeft = mRectSize / 2
        mPanelMarginTop = (h - w) / 2 + mPanelMarginLeft
        mTextSize = mRectSize - rectPadding * 2
        mSourcePaint.textSize = mTextSize.toFloat()
        mInputPaint.textSize = mTextSize.toFloat()

        val fontMetrics = mSourcePaint.fontMetricsInt
        mTextBaseline = (mRectSize - fontMetrics.bottom + fontMetrics.top) / 2 - fontMetrics.top

        val noteFontMetrics = mTipsPaint.fontMetricsInt
        for (i in mNoteTextBaseline.indices) {
            mNoteTextBaseline[i] = (mRectSize / 3 * (i * 2 + 1) - noteFontMetrics.bottom + noteFontMetrics.top) / 2 - noteFontMetrics.top
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val width = canvas.width
        val height = canvas.height

        for (i in 0 until SudokuPresenter.GRID_COUNT) {
            val column = i % 9
            val row = i / 9

            val gridBean = mSudokuPresenter.getGrid(i)
            if (gridBean.source != 0) {
                drawSource(canvas, gridBean.source, column, row)
            } else if (gridBean.input != 0) {
                drawInput(canvas, gridBean.input, column, row)
            } else {
                drawTips(canvas, gridBean.tips, column, row)
            }
        }
        for (i in 0..9) {
            canvas.drawLine(mPanelMarginLeft.toFloat(), (mPanelMarginTop + mRectSize * i).toFloat(),
                    (width - mPanelMarginLeft).toFloat(), (mPanelMarginTop + mRectSize * i).toFloat(),
                    if (i % 3 == 0) mBoldLinePaint else mThinLinePaint)
        }
        for (i in 0..9) {
            canvas.drawLine((mPanelMarginLeft + mRectSize * i).toFloat(), mPanelMarginTop.toFloat(),
                    (mPanelMarginLeft + mRectSize * i).toFloat(), (height - mPanelMarginTop).toFloat(),
                    if (i % 3 == 0) mBoldLinePaint else mThinLinePaint)
        }
    }

    /**
     * 绘制原始的数字
     */
    private fun drawSource(canvas: Canvas, num: Int, column: Int, row: Int) {
        canvas.drawText(num.toString(),
                (mPanelMarginLeft + column * mRectSize + mRectSize / 2).toFloat(),
                (mPanelMarginTop + mTextBaseline + row * mRectSize).toFloat(),
                mSourcePaint)
    }

    /**
     * 绘制输入的数字
     */
    private fun drawInput(canvas: Canvas, num: Int, column: Int, row: Int) {
        canvas.drawText(num.toString(),
                (mPanelMarginLeft + column * mRectSize + mRectSize / 2).toFloat(),
                (mPanelMarginTop + mTextBaseline + row * mRectSize).toFloat(),
                mInputPaint)
    }

    /**
     * 绘制提示数字
     */
    private fun drawTips(canvas: Canvas, tips: IntArray, column: Int, row: Int) {
        for (j in 0..8) {
            if (tips[j] == 0) {
                continue
            }
            canvas.drawText((j + 1).toString(),
                    (mPanelMarginLeft + column * mRectSize + mRectSize / 6 * (j % 3 * 2 + 1)).toFloat(),
                    (mPanelMarginTop + mNoteTextBaseline[j / 3] + row * mRectSize).toFloat(),
                    mTipsPaint)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (mSudokuPresenter.checkInput()) {
            mSudokuPresenter.reset()
        } else {
            mSudokuPresenter.updateTips()
            if (mSudokuPresenter.checkInput()) {
                Toast.makeText(context, "Suc", Toast.LENGTH_SHORT).show()
            }
        }
        postInvalidate()
        return super.onTouchEvent(event)
    }
}
